client/httplib/httpclient.go

```diff
@@ -0,0 +1,296 @@
+	pointer *responsePointer
+}
+
+type responsePointer struct {
+	response      **http.Response
+	statusCode    **int
+	header        **http.Header
+	headerValues  map[string]**string //用户传一个key，然后将key存在map的key里，header的value存在value里
+	contentLength **int64
+}
+
+// NewClient return a new http client
+func NewClient(name string, endpoint string, opts ...ClientOption) (*Client, error) {
+	res := &Client{
+		Name:     name,
+		Endpoint: endpoint,
+	}
+	setting := GetDefaultSetting()
+	res.Setting = &setting
+	for _, o := range opts {
+		err := o(res)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+// Response will set response to the pointer
+func (c *Client) Response(resp **http.Response) *Client {
+	if c.pointer == nil {
+		newC := *c
+		newC.pointer = &responsePointer{
+			response: resp,
+		}
+		return &newC
+	}
+	c.pointer.response = resp
+	return c
+}
+
+// StatusCode will set response StatusCode to the pointer
+func (c *Client) StatusCode(code **int) *Client {
+	if c.pointer == nil {
+		newC := *c
+		newC.pointer = &responsePointer{
+			statusCode: code,
+		}
+		return &newC
+	}
+	c.pointer.statusCode = code
+	return c
+}
+
+// Headers will set response Headers to the pointer
+func (c *Client) Headers(headers **http.Header) *Client {
+	if c.pointer == nil {
+		newC := *c
+		newC.pointer = &responsePointer{
+			header: headers,
+		}
+		return &newC
+	}
+	c.pointer.header = headers
+	return c
+}
+
+// HeaderValue will set response HeaderValue to the pointer
+func (c *Client) HeaderValue(key string, value **string) *Client {
+	if c.pointer == nil {
+		newC := *c
+		newC.pointer = &responsePointer{
+			headerValues: map[string]**string{
+				key: value,
+			},
+		}
+		return &newC
+	}
+	if c.pointer.headerValues == nil {
+		c.pointer.headerValues = map[string]**string{}
+	}
+	c.pointer.headerValues[key] = value
+	return c
+}
+
+// ContentType will set response ContentType to the pointer
+func (c *Client) ContentType(contentType **string) *Client {
+	return c.HeaderValue("Content-Type", contentType)
+}
+
+// ContentLength will set response ContentLength to the pointer
+func (c *Client) ContentLength(contentLength **int64) *Client {
+	if c.pointer == nil {
+		newC := *c
+		newC.pointer = &responsePointer{
+			contentLength: contentLength,
+		}
+		return &newC
+	}
+	c.pointer.contentLength = contentLength
+	return c
+}
+
+// setPointers set the http response value to pointer
+func (c *Client) setPointers(resp *http.Response) {
+	if c.pointer == nil {
+		return
+	}
+	if c.pointer.response != nil {
+		*c.pointer.response = resp
+	}
+	if c.pointer.statusCode != nil {
+		*c.pointer.statusCode = &resp.StatusCode
+	}
+	if c.pointer.header != nil {
+		*c.pointer.header = &resp.Header
+	}
+	if c.pointer.headerValues != nil {
+		for k, v := range c.pointer.headerValues {
+			s := resp.Header.Get(k)
+			*v = &s
+		}
+	}
+	if c.pointer.contentLength != nil {
+		*c.pointer.contentLength = &resp.ContentLength
+	}
+}
```

Comment:
See my above comments. `pointer` is not thread-safe

Reply:
maybe we can alway return new Client struct point. like

```
func (c *Client) ContentLength(contentLength **int64) *Client {
        newC := *c
	if c.pointer == nil {
               // setting or pointer change to value type
		newC.pointer = responsePointer{
			contentLength: contentLength,
		}
		return &newC
	}
	c.pointer.contentLength = contentLength
	return &newC
}
```

Reply:
I think it's not a good idea. From the original design, creating one client instance for one downstream. For example, when users need to communicate with two different severs, they only need to create two client instances. And users could reuse them everywhere. 

Client should be "immutable", which means that users should not change client after creating it. 

So it's easier for users to use client. If we need to create clients every time, why users do not use `BeegoHttpRequest` directly?

Reply:
Or can we wrapper request and save the point in new request struct？In this way, Client is also immutable, and  chain invocation also can be done. And it also thread safe, cause every request has one and only request struct instance.

```go
// when client first call contenLen or Header, will create a new request
func (c *Client) ContentLength(contentLength **int64) *req{
	return c.newRequest(c, WithContentLength(contentLength ))
}

// new request struct
type req struct {
  client client  // save current client setting
  opts []resOptions
}

// and new req also have same function signature to keep chain go on
func (r *req) ContentLength(contentLength **int64) *req{
        r.opts = append(r.opts, WithContentLength(contentLength ))
	return r
}
```